src/ValueObject/Request/Request.php

```diff
@@ -0,0 +1,165 @@
+    private $method;
+
+    /** @var string[] */
+    private $headers;
+
+    /** @var string */
+    private $body;
+
+    /**
+     * Request constructor.
+     *
+     * @param string        $uri
+     * @param string        $method
+     * @param string[]      $headers
+     * @param string | null $body
+     */
+    public function __construct($uri, $method, $headers = array(), $body = null)
+    {
+        $this->uri = (string) $uri;
+        $this->method = (string) $method;
+
+        $this->headers = [];
+        foreach ($headers as $key => $header) {
+            $this->headers[(string) $key] = (string) $header;
+        }
+
+        // read body from php input stream, if no body was set
+        $this->body = (string) $body;
+        if (empty($body)) {
+            $this->body = file_get_contents('php://input');
+        }
+    }
+
+    /**
+     * @return string
+     */
+    public function getUri()
+    {
+        return $this->uri;
+    }
+
+    /**
+     * @return \string[]
+     */
+    public function getHeaders()
+    {
+        return $this->headers;
+    }
+
+    /**
+     * @return string
+     */
+    public function getMethod()
+    {
+        return $this->method;
+    }
+
+    /**
+     * @param string $headerKey
+     *
+     * @return string | null
+     */
+    public function getHeader($headerKey)
+    {
+        return (empty($this->headers[$headerKey])
+            ? null
+            : ((string) $this->headers[$headerKey])
+        );
+    }
+
+    /**
+     * @return bool|string
+     */
+    public function getBody()
+    {
+        return $this->body;
+    }
+
+    /**
+     * @param Request $request
+     * @param string  $paramName
+     *
+     * @return string | null
+     *
+     * @throws Exception\BadRequest
+     */
+    public function getParam($paramName)
+    {
+        $data = [];
+        parse_str(parse_url($this->getUri(), PHP_URL_QUERY), $data);
+
+        // parse either from request query or from request body
+        if (!empty($data[$paramName])) {
+            return $data[$paramName];
+        } else {
+            $contentTypeKey = 'Content-Type';
+            $requestHeaders = $this->getHeaders();
+
+            // check if there is a content type header provided
+            if (empty($requestHeaders[$contentTypeKey])) {
+                throw new Exception\BadRequest('Invalid request body.');
+            }
+
+            // check if the provided content type is supported
+            switch($this->parseContentType($requestHeaders[$contentTypeKey])) {
+                case 'application/json':
+                    $data = json_decode($this->getBody(), true);
+                    if (!empty($data[$paramName])) {
+                        return $data[$paramName];
+                    }
+                    break;
+                case 'application/x-www-form-urlencoded':
+                    parse_str($this->getBody(), $data);
+                    if (!empty($data[$paramName])) {
```

Comment:
since this logic duplicates can move this logic outside of switch into the final return: ```return !empty($data[$paramName]) ? $data[$paramName] : null```

Reply:
Thanks. I could even include the first return statement within the if block and loose two more lines that way.